### PR TITLE
Document invariants of raptyMiddle

### PR DESCRIPTION
**Pull request by Lysxia**
Also a little refactoring

- renaming `xs''FT` and `fs` to `midxs`
- renaming `firstf`, `lastf` to `firstx`, `lastx`
- renaming `s` to `lenys`
- regroup the arguments between (1) the constants, (2) those that get updated uniformly at every recursive call (representing powers of 3), and (3) the non-uniform ones
- shift the logic around `sizec` "one level up", so the new `sizec` is the old `3 * sizec` also equal to the old `sizec'`.

### CODE REVIEW

File: containers/src/Data/Sequence/Internal.hs [hunk 1]
```diff
@@ -1461,97 +1461,104 @@ before3FT firstx xs lastx =
...
-           fs
-           (RCountMid s pr 1 sf)
-     = deep
-            (Two pr (Node3 (3 * sizec) firstf firstf firstf))
-            (mapMulFT s swizzle fs)
-            (Two (Node3 (3 * sizec) lastf lastf lastf) sf)
+raptyMiddle midxs lenys
+            !sizec
+            firstf
+            lastf
```
Comment (treeowl) [inline containers/src/Data/Sequence/Internal.hs]: We should probably rename `firstf -> firstx` and `lastf -> lastx`. Could you amend that to this commit?

**Review by treeowl — CHANGES_REQUESTED** [review]
Definitely an improvement. One change requested before I can merge to my branch. I also have to double check that the argument regrouping doesn't muck up unboxing—GHC can get a bit fussy when there are lots of arguments.

**Comment by treeowl** [comment]
I've squashed these and added them (as a single commit) to my branch. Many thanks, @Lysxia! I have a couple questions:

1. Do you prefer to be called Li-Yao Xia or Xia Li-Yao? I see one way here and another on StackOverflow. Also, for my personal curiosity, which name is your personal name and which is a family name (or patronymic?)?
2. How would you feel about similarly documenting the (presumably somewhat simpler) invariants in `aptyMiddle`, while you have these ideas running around in your head? I've struggled for years to try to hold all the ideas there together at any one time.

**Comment by Lysxia** [comment]
Great!

1. Li-yao is my first name, so you can credit me as Li-yao Xia. (Chinese people commonly call each other by family-name-given-name so I'm used to both ways.)

2. Sure!

**Comment by treeowl** [comment]
Thanks a lot!

**Pull request closed without merging**